Use BasicURLHandler instead of okhttp variant to support newer SBT versions

robobario opened the pull request:
When trying to `whPublish` a project using a newer SBT version like 1.9.3 we see exceptions like:
```
[error] java.lang.NoClassDefFoundError: sbt/internal/librarymanagement/ivyint/GigahorseUrlHandler
[error]     at org.jboss.pnc.sbt.plugins.PublishWithHeadersPlugin$$anon$1.<init>(PublishWithHeadersPlugin.scala:37)
[error]     at org.jboss.pnc.sbt.plugins.PublishWithHeadersPlugin$.$anonfun$projectSettings$3(PublishWithHeadersPlugin.scala:36)
[error]     at org.jboss.pnc.sbt.plugins.PublishWithHeadersPlugin$.$anonfun$projectSettings$3$adapted(PublishWithHeadersPlugin.scala:33)
```

`okhttp` and the `GigahorseUrlHandler` were removed from [SBT in 1.7.0](https://github.com/sbt/librarymanagement/commit/76452e53ff6f7211db2bab5dd3bc8f06dac1bbce) for security reasons and instead they use an BasicURLHandler based on URLConnection.

Downgrading SBT is no longer a simple fix when trying to build scala 2.13.11+ as the project has some tight integration with SBT, so it makes sense to make the plugin compatible with newer SBT versions.

This change updates the plugin to be based on the `BasicURLHandler` (unfortunately a lot of the private guts of the class had to be copied and switched to scala to keep the same behaviour, just adding the extra headers in)

I tested that it worked with a local nexus repository and observed the headers being set on PUT requests.

robobario committed:
Use ivy BasicURLHandler instead of okhttp variant

Why:
Okhttp was removed from SBT in 1.7 for security reasons. They switched
to using the ivy BasicURLHandler instead. This updates the plugin to
extend that class (sadly having to duplicate much of the innards).

This is required to build projects that require SBT >1.7, for example
0.0.1 fails when trying to use it to build scala 2.13.11 or 2.13.12
due to the missing class and downgrading SBT is not feasible as the
project has complex SBT integration.

mimaison commented on src/main/scala/org/jboss/pnc/sbt/plugins/WithHeadersURLHandler.scala:
```diff
@@ -18,60 +18,160 @@
-    Message.debug("Uploading using WithHeadersURLHandler...")
+    if (!("http" == dest.getProtocol) && !("https" == dest.getProtocol)) throw new UnsupportedOperationException("URL repository only support HTTP PUT at the moment")
 
+    // Install the IvyAuthenticator// Install the IvyAuthenticator
```
This comment is duplicated.
The previous code was also installing `ErrorMessageAuthenticator`, I think we should keep it as it may help debug authentication issues.

robobario replied:
will rm the comment altogether, the method is perfectly understandable `IvyAuthenticator.install()`

mimaison commented on src/main/scala/org/jboss/pnc/sbt/plugins/WithHeadersURLHandler.scala:
```diff
@@ -18,60 +18,160 @@
+        builder.append(String.join("\n    ", header.getValue))
+        builder.append("\n")
+      }
+    }
```
I think (not tested) we can use:
```
headers.entrySet().forEach(entry => {
    val key: String = entry.getKey
    if (key != null) {
        builder.append(key)
        builder.append(": ")
    }
    builder.append(String.join("\n    ", entry.getValue))
    builder.append("\n")
})
```
to avoid using the deprecated `JavaConversions` class

mimaison commented on src/main/scala/org/jboss/pnc/sbt/plugins/WithHeadersURLHandler.scala:
```diff
@@ -18,60 +18,160 @@
+import java.util
 
-class WithHeadersURLHandler(headers: Headers) extends GigahorseUrlHandler(URLHandlerHelper.http) {
+class WithHeadersURLHandler(headers: java.util.Map[String,String]) extends BasicURLHandler() {
```
It's a matter of style, but to be consistent with the code below, we can have `util.Map` instead of `java.util.Map`

mimaison commented on project/build.properties:
```diff
@@ -1 +1 @@
-sbt.version = 1.3.12
+sbt.version = 1.9.3
```
What about bumping this to 1.9.9 (the current latest version)? Although this project is probably not affected, it seems there were a couple of security fixes in the last few releases.

mimaison commented on src/main/scala/org/jboss/pnc/sbt/plugins/WithHeadersURLHandler.scala:
```diff
@@ -18,60 +18,160 @@
-      .headers(headers)
-      .put(body)
-      .build()
+  @throws[IOException]
```
I understand you converted the Java code from `BasicURLHandler.upload()` which has `throws IOException`. Since this method does not throw IOException, maybe this could be dropped.

mimaison reviewed (approved):
Thanks for the PR! The changes look good to me, I just left a few suggestions.

robobario committed:
Apply review feedback:

- install ErrorMessageAuthenticator
- java.util.Map -> util.Map
- remove incorrect @throws declaration
- avoid JavaConversions
- reformat

robobario committed:
Upgrade SBT to latest

matejonnet commented:
Thanks for providing the update.
